core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -556,6 +562,42 @@ public Object readGenericValue() throws IOException {
         return null;
     }
 
+    /**
+     * Reads a {@link NamedWriteable} from the current stream, by first reading its name and then looking for
+     * the corresponding entry in the registry by name, so that the proper object can be read and returned.
+     */
+    <C> C readNamedWriteable(String category) throws IOException {
+        String name = readString();
+        NamedWriteable<C> namedWriteable = namedWriteableRegistry.getPrototype(category, name);
+        return namedWriteable.readFrom(this);
```

Comment:
I will give this a shot. What feels a bit weird is that StreamInput would have methods to support NamedWriteables but it really doesn't support them as it doesn't have a registry, only a subclass will. Will play around with this though and see what it looks like after the change.